### PR TITLE
Make polynomial regression more robust

### PR DESCRIPTION
**Pull request by keskitalo**
Existing implementation of the polynomial filter is fragile. It was using SVD to overcome singular template covariance matrix and applied an arbitrary rcond cut to small singular values. When small singular values were encountered, the results were compromised.

The new implementation accepts singular values down to machine precision and does away with the squaring of the template matrix. The latter greatly improved the condition number of the system.

### CODE REVIEW

File: src/libtoast/src/toast_tod_filter.cpp [hunk 1]
```diff
@@ -161,6 +134,11 @@ void toast::filter_polynomial(int64_t order, size_t n, uint8_t * flags,
                 }
             }
         }
+
+        if (info != 0) {
+            std::cerr << "ERROR: DGELLS info = " << info << std::endl;
+        }
```
Comment (tskisner) [inline src/libtoast/src/toast_tod_filter.cpp:140]: Do we want to remove this debugging statement?  Or use the toast logging to make an error message and raise an exception pointing to the line number?  Like this:
```
if (info != 0) {
    auto log = toast::Logger::get();
    std::ostringstream o;
    o << "DGELLS:  " << ngood << "/" << scanlen << " good samples, order " << norder;
    o << " failed with info " << info;
    log.error(o.str().c_str(), TOAST_HERE());
    throw std::runtime_error(o.str().c_str());
}
```

File: src/libtoast/src/toast_tod_filter.cpp [hunk 2]
```diff
@@ -34,7 +34,7 @@ void toast::filter_polynomial(int64_t order, size_t n, uint8_t * flags,
     double fone = 1.0;
 
     #pragma \
-    omp parallel for schedule(static) default(none) shared(order, signals, flags, n, nsignal, starts, stops, nscan, norder, upper, lower, notrans, trans, fzero, fone)
```
Comment (tskisner) [inline src/libtoast/src/toast_tod_filter.cpp:37]: Maybe use default(shared)?

**Review by tskisner — APPROVED** [review]
Thanks for this cleanup / testing.

File: src/libtoast/src/toast_tod_filter.cpp (hunk 1 above)
Comment (keskitalo) [reply src/libtoast/src/toast_tod_filter.cpp]: logging and throwing is much better. I adopted that.

File: src/libtoast/src/toast_tod_filter.cpp (hunk 2 above)
Comment (keskitalo) [reply src/libtoast/src/toast_tod_filter.cpp]: I prefer not to use defaults. This way every shared variable is an explicit choice. Fewer surprises that way.

Comment (tskisner) [reply src/libtoast/src/toast_tod_filter.cpp]: But that is exactly what `default(none)` (now removed) does.  If no `default` is specified, then all unspecified variables are shared.  [One reference](https://rookiehpc.org/openmp/docs/default/index.html) and [another](https://www.openmp.org/spec-html/5.2/openmpsu35.html).

Comment (keskitalo) [reply src/libtoast/src/toast_tod_filter.cpp]: Correct. I removed default(none) temporarily during development and forgot to add it back. It is now restored.

**Pull request merged**